### PR TITLE
Fixed invalid prop types error message to be more specific

### PR DESCRIPTION
**Pull request by NicBonetto**
**Before submitting a pull request,** please make sure the following is done:

1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
2. Run `yarn` in the repository root.
3. If you've fixed a bug or added code that should be tested, add tests!
4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
5. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
6. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
7. Run the [Flow](https://flowtype.org/) typechecks (`yarn flow`).
8. If you haven't already, complete the CLA.

**Learn more about contributing:** https://reactjs.org/docs/how-to-contribute.html

### CODE REVIEW

**Comment by NicBonetto** [comment]
#11289

File: packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js [hunk 1]
```diff
@@ -183,10 +183,12 @@ if (__DEV__) {
     if (typeof value === 'boolean') {
       warning(
         DOMProperty.shouldAttributeAcceptBooleanValue(name),
-        'Received `%s` for non-boolean attribute `%s`. If this is expected, cast ' +
-          'the value to a string.%s',
+        'If you intentionally tried to pass a boolean, pass it as a string ' +
+          'instead: `%s`="`%s`". If you mean to conditionally pass an ' +
+          'attribute, use a ternary expression: `%s`={condition ? value : null}.%s',
```
Comment (sebmarkbage) [inline packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: If the value is `true` then this is very unlikely to happen because the problematic pattern is `condition && realValue` specifically. I can't think of a way where you'd end up with `true` for the same reason. Maybe we should have two separate error messages for `true` and `false` where this recommendation is only for the `false` case?

Maybe also call out that pattern specifically, so that people know what's wrong and what to look for.

Also, the recommendation should be to pass `undefined`, not `null`. It doesn't really matter for DOM components, but if it is a custom component that is the difference between defaultProps working and not working. So it's better to have consistent advice.

**Comment by NicBonetto** [comment]
I may need some help with the error that occurred in the integration test. I believe it is coming from adding an additional condition in the `if` statement, and I would appreciate a little direction. Thank you.

File: packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js [hunk 2]
```diff
@@ -189,8 +200,6 @@ if (__DEV__) {
         name,
         getStackAddendum(),
       );
-      warnedProperties[name] = true;
-      return true;
```
Comment (gaearon) [inline packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: Why was this removed?

File: packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js [hunk 3]
```diff
@@ -180,26 +180,28 @@ if (__DEV__) {
...
-        'If you mean to conditionally pass an attribute, use a ternary ' +
-          'expression: `%s`={condition ? value : undefined} instead of ' +
-          '{condition && value}.%s',
-        value,
-        getStackAddendum(),
-      );
+    if (typeof value === 'boolean') {
+      if (value === true) {
+        warning(
+          DOMProperty.shouldAttributeAcceptBooleanValue(name),
```
Comment (gaearon) [inline packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: Can we make this `warning(false` and move out this (negated) condition together with the top one?

Comment (NicBonetto) [reply packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: Yes :) Just so I have a correct understanding, the refactor will look like:
```
if (typeof value === 'boolean' && !false) {
  if (value === true) {
    warning(false,
      ...
    );
  } else {
    warning(false,
      ...
    );
  }
  ...
}
```
Is this the branching logic you prefer?

Comment (gaearon) [reply packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]:
```js
if (typeof value === 'boolean' && !DOMProperty.shouldAttributeAcceptBooleanValue(name)) {
  if (value === true) {
    warning(false,
      ...
    );
  } else {
    warning(false,
      ...
    );
  }
  ...
}

File: packages/react-dom/src/__tests__/ReactDOMAttribute-test.js [hunk 4]
```diff
@@ -54,8 +54,8 @@ describe('ReactDOM unknown attribute', () => {
       expectDev(
         normalizeCodeLocInfo(console.error.calls.argsFor(0)[0]),
       ).toMatch(
-        'Warning: Received `true` for non-boolean attribute `unknown`. ' +
-          'If this is expected, cast the value to a string.\n' +
+        'Warning: Received `true` for non-boolean attribute `unknown`. If this is expected, cast ' +
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMAttribute-test.js]: This change is not unnecessary, right? It just reformats the message in a slightly less readable way.

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 5]
```diff
@@ -2042,7 +2042,8 @@ describe('ReactDOMComponent', () => {
       expect(el.hasAttribute('whatever')).toBe(false);
 
       expectDev(console.error.calls.argsFor(0)[0]).toContain(
-        'Warning: Received `true` for non-boolean attribute `whatever`',
+        'Received `true` for non-boolean attribute `whatever`. If this is expected, cast ' +
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Would look nicer if `If this is expected...` started on the new line.

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 6]
```diff
@@ -2055,7 +2056,8 @@ describe('ReactDOMComponent', () => {
...
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Same

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 7]
```diff
@@ -2230,7 +2232,8 @@ describe('ReactDOMComponent', () => {
       expect(el.hasAttribute('whatever')).toBe(false);
 
       expectDev(console.error.calls.argsFor(0)[0]).toContain(
-        'Warning: Received `true` for non-boolean attribute `whatever`.',
+        'Received `true` for non-boolean attribute `whatever`. If this is expected, cast ' +
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Same

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 8]
```diff
@@ -2283,7 +2286,9 @@ describe('ReactDOMComponent', () => {
 
       expectDev(console.error.calls.count()).toBe(1);
       expectDev(console.error.calls.argsFor(0)[0]).toContain(
-        'Warning: Received `false` for non-boolean attribute `whatever`.',
+        'If you mean to conditionally pass an attribute, use a ternary ' +
+          'expression: `false`={condition ? value : undefined} instead of ' +
+          '{condition && value}.',
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: What does `false={condition? value : undefined}` mean? Why does it says `false`?

Please fix this to show the attribute name.

File: packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js [hunk 9]
```diff
@@ -180,15 +180,29 @@ if (__DEV__) {
...
+          'Received `%s` for non-boolean attribute `%s`. If this is expected, cast ' +
+            'the value to a string.%s',
+          value,
+          name,
+          getStackAddendum(),
+        );
+      } else {
+        warning(
+          false,
+          'If you mean to conditionally pass an attribute, use a ternary ' +
```
Comment (gaearon) [inline packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: Can this also start with `Received ... for non-boolean attribute ...`? I want both messages to look similar, but with different second sentences.

**Comment by gaearon** [comment]
I tried to build React locally and then used

```js
      ReactDOM.render(
        <h1 myattribute={false}>Hello World!</h1>,
        document.getElementById('container')
      );
```

I saw this message:

screen shot 2017-10-28 at 6 50 25 pm

Does it make sense to you? I don't think its says what it's supposed to say.

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 10]
```diff
@@ -2283,7 +2286,9 @@ describe('ReactDOMComponent', () => {
 
       expectDev(console.error.calls.count()).toBe(1);
       expectDev(console.error.calls.argsFor(0)[0]).toContain(
-        'Warning: Received `false` for non-boolean attribute `whatever`.',
+        'Received `false` for non-boolean attribute `whatever`. If you mean to conditionally ' +
+          'pass an attribute, use a ternary expression: {`whatever` ? value : undefined} ' +
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Please make sure to read the messages you assert in tests. It looks like you were trying to adjust tests to match the wrong behavior in the code, rather than the other way around.

Comment (NicBonetto) [reply packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Sorry, my understanding was that the value being passed was the condition. So `value` should hold the attribute name instead of `condition` correct? It didn't make sense that `whatever` (holding the value `false`) could be an attribute.

Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: I’m sorry that I snapped. I had a bad day and shouldn’t have spoken to you with this tone. I apologize.

In this test, `whatever` is the *attribute name*. Not the value. You can see if from the first sentence:

>Received `false` for non-boolean attribute `whatever`.

Therefore, a condition like `{`whatever` ? value : undefined}` is not relevant to the user code. The user has something like `whatever={condition && value}` with `condition` being `false`. We want to tell the user to turn it into something like `whatever={condition ? value : undefined}`.

So I think ideally the messages would be:

1. If you get `true`:

```js
expectDev(console.error.calls.argsFor(0)[0]).toContain(
  'Received `true` for non-boolean attribute `whatever`.\n\n' +
  'If you want `true` to appear in the DOM, cast it to a string. ' +
  'For example, you can pass `whatever="true"` or `whatever={String(value)}` instead.'
);
```

2. If you get `false`:

```js
expectDev(console.error.calls.argsFor(0)[0]).toContain(
  'Received `false` for non-boolean attribute `whatever`.\n\n' +
  'If you want `false` to appear in the DOM, cast it to a string. ' +
  'For example, you can pass `whatever="false"` or `whatever={String(value)}`.\n\n' +
  'If you want to omit `whatever` based on a condition, use a ternary expression. ' +
  'For example, you can pass `whatever={condition ? value : undefined}` ' +
  'instead of `whatever={condition && value}`.'
);
```

I hope this is helpful and clarifies things!

Comment (NicBonetto) [reply packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: No need to apologize, your message did not come off badly. I only apologized for making you revise this pr so many times because I did not grasp the whole picture. Yes, this is extremely helpful. Thank you.

Comment (NicBonetto) [reply packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: Do you prefer:
```
`whatever={String(value)}`
```
or
```
`whatever={value.toString()}`
```

Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMComponent-test.js]: I thought about the second one but it will throw for null/undefined. Might be unexpected. First one seems safer. Although on the other hand you probably don't want `"undefined"` or `"null"` as strings either.

I wish there was some concise way to express "stringify unless it's null/undefined".

I guess I like that `value.toString()` throws it it doesn't exist. If you mess it up you'll immediately realize it by the crash, instead of quietly using the wrong value. So let's suggest that.

File: packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js [hunk 11]
```diff
@@ -180,15 +180,45 @@ if (__DEV__) {
...
-          'the value to a string.%s',
-        value,
-        name,
-        getStackAddendum(),
-      );
+    if (
+      typeof value === 'boolean' &&
+      !DOMProperty.shouldAttributeAcceptBooleanValue(name)
+    ) {
+      if (value === true) {
```
Comment (gaearon) [inline packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js]: Nit: we already know it's a boolean so can just write `if (value)`

**Comment by gaearon** [comment]
Thanks for the fixes!

I looked at it in practice, and I think it's a bit too noisy with backticks everywhere. Let's keep them around singular words (e.g. around attribute name), but remove them around fragments of code.

I also found the current messages too noisy and potentially misleading after all. Sorry about this—it takes a few iterations to get this right. Here's my new proposal.

For `true`:

```js
expectDev(console.error.calls.argsFor(0)[0]).toContain(
  'Received `true` for non-boolean attribute `whatever`.\n\n' +
  'If you want to write it to the DOM, pass a string instead: ' +
  'whatever="true" or whatever={value.toString()}.'
);
```

screen shot 2017-10-29 at 17 51 51


For `false`:

```js
expectDev(console.error.calls.argsFor(0)[0]).toContain(
  'Received `false` for non-boolean attribute `whatever`.\n\n' +
  'If you want to write it to the DOM, pass a string instead: ' +
  'whatever="false" or whatever={value.toString()}.\n\n' +
  'If you conditionally omit it with whatever={condition && value}, ' +
  'pass whatever={condition ? value : undefined} instead.'
);
```

screen shot 2017-10-29 at 17 51 42

I hope it’s not too much of a burden! I think these are clearer.

**Comment by gaearon** [comment]
I think this is good to go. Did a few minor changes. Thanks so much!
